### ISSUE TITLE
Error for inconsistent types in ternary in 1.3.4

### ISSUE THREAD

**Issue by grimm26** [hashicorp/terraform#32200]
### Terraform Version

```shell
Terraform v1.3.4
on linux_amd64
```

### Terraform Configuration Files

```terraform
variable "object_lifecycle_preset" {
  description = <<-EOL
  Choose a pre-configured set of object lifecycle rules for your bucket.
  - empty_bucket: Delete all objects from the bucket. Use this in preparation to destroy the bucket. This is non-recoverable.
  EOL
  type        = string
  default     = "default"
  validation {
    condition     = contains(["empty_bucket", "default"], var.object_lifecycle_preset)
    error_message = "You must select from one of the available presets: empty_bucket."
  }
}

variable "lifecycle_object_rules" {
  description = "List of object describing object lifecycle rules on the noncurrent versions of objects."
  type = list(object(
    {
      id                                     = string
      enabled                                = bool
      prefix                                 = optional(string)
      tags                                   = optional(map(string), {})
      object_size_greater_than               = optional(number)
      object_size_less_than                  = optional(number)
      abort_incomplete_multipart_upload_days = optional(number)
      transitions                            = optional(map(string), {})
      noncurrent_transitions                 = optional(map(string), {})
    }
  ))
  default = []
}

locals {
  lifecycle_presets = {
    default = [
      {
        id = "default noncurrent transition"
        # prefix = "foo/bar/"
        # tags = { "key" = "value", "key2" = "value2" }
        enabled = true
        # transitions = {
        #   standard_ia = 120
        # }
        object_size_greater_than = 1024 * 128
        noncurrent_transitions = {
          standard_ia = 30
        }
      },
      {
        id                                     = "default noncurrent expiration"
        enabled                                = true
        abort_incomplete_multipart_upload_days = 1
        noncurrent_transitions = {
          expiration = 120
        }
        transitions = {
          expired_object_delete_marker = true
        }
      },
    ]
    empty_bucket = [
      {
        id      = "empty bucket1"
        enabled = true
        noncurrent_transitions = {
          expiration = 1
        }
        transitions = {
          expiration = 1
        }
        abort_incomplete_multipart_upload_days = 1
      },
      {
        id      = "empty bucket2"
        enabled = true
        transitions = {
          expired_object_delete_marker = true
        }
      },
    ]
  }

  # If we have custom rules specified, use those. Otherwise use the preset.
  lifecycle_object_rules = length(var.lifecycle_object_rules) > 0 ? var.lifecycle_object_rules : local.lifecycle_presets[var.object_lifecycle_preset]
}
```

### Debug Output

https://gist.github.com/013226b0be1988e7e5fcfe97f2d53abb

### Expected Behavior

no error

### Actual Behavior

```

│ Error: Inconsistent conditional result types
│ 
│   on main.tf line 83, in locals:
│   83:   lifecycle_object_rules = length(var.lifecycle_object_rules) > 0 ? var.lifecycle_object_rules : local.lifecycle_presets[var.object_lifecycle_preset]
│     ├────────────────
│     │ local.lifecycle_presets is object with 2 attributes
│     │ var.lifecycle_object_rules is empty list of object
│     │ var.object_lifecycle_preset is "default"
│ 
│ The true and false result expressions must have consistent types. The 'true' value is list of object, but the 'false' value is tuple.
```

### Steps to Reproduce

1. use terraform 1.3.4
2. terraform init
3. terraform plan

### Additional Context

This works in terraform 1.3.3

**Comment by 42Atomys** [comment]
I have the similar issue with 1.3.4 (works in 1.3.3) for `inconsistent map element types`

Following the stacktrace : 

<details><summary>Show Stacktrace</summary>

```
inconsistent map element types (cty.Object(map[string]cty.Type{"extraServers":cty.List(cty.ObjectWithOptionalAttrs(map[string]cty.Type{"hosts":cty.List(cty.String), "port":cty.Object(map[string]cty.Type{"name":cty.String, "number":cty.Number, "protocol":cty.String}), "tls":cty.ObjectWithOptionalAttrs(map[string]cty.Type{"credentialName":cty.String, "httpsRedirect":cty.Bool, "mode":cty.String}, []string{"credentialName", "httpsRedirect", "mode"})}, []string{"tls"})), "hosts":cty.List(cty.String), "ingressSelectorName":cty.String, "namespace":cty.String, "serverHttpsRedirect":cty.Bool, "tlsCredentialName":cty.String, "tlsMode":cty.String}) then cty.Object(map[string]cty.Type{"extraServers":cty.List(cty.Object(map[string]cty.Type{"hosts":cty.List(cty.String), "port":cty.Object(map[string]cty.Type{"name":cty.String, "number":cty.Number, "protocol":cty.String}), "tls":cty.Object(map[string]cty.Type{"credentialName":cty.String, "httpsRedirect":cty.Bool, "mode":cty.String})})), "hosts":cty.List(cty.String), "ingressSelectorName":cty.String, "namespace":cty.String, "serverHttpsRedirect":cty.Bool, "tlsCredentialName":cty.String, "tlsMode":cty.String}))
goroutine 190 [running]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65
runtime/debug.PrintStack()
        /usr/local/go/src/runtime/debug/stack.go:16 +0x19
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        /home/circleci/project/project/internal/logging/panic.go:55 +0x153
panic({0x21e3360, 0xc001196610})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/zclconf/go-cty/cty.MapVal(0xc000ddcc50)
        /home/circleci/go/pkg/mod/github.com/zclconf/go-cty@v1.12.0/cty/value_init.go:221 +0x47f
github.com/zclconf/go-cty/cty.transform({0x0?, 0x0, 0x0}, {{{0x2b6c868?, 0xc0005fb780?}}, {0x221b780?, 0xc000dd4750?}}, {0x2b53c00, 0xc00013af70})
        /home/circleci/go/pkg/mod/github.com/zclconf/go-cty@v1.12.0/cty/walk.go:204 +0xc3e
github.com/zclconf/go-cty/cty.TransformWithTransformer(...)
        /home/circleci/go/pkg/mod/github.com/zclconf/go-cty@v1.12.0/cty/walk.go:130
github.com/hashicorp/hcl/v2/ext/typeexpr.(*Defaults).Apply(0xc0008814c0?, {{{0x2b6c868?, 0xc0005fb780?}}, {0x221b780?, 0xc000dd4750?}})
        /home/circleci/go/pkg/mod/github.com/hashicorp/hcl/v2@v2.14.1/ext/typeexpr/defaults.go:38 +0xa7
github.com/hashicorp/terraform/internal/terraform.prepareFinalInputVariableValue({{0xc00095f8e0, 0x1, 0x1}, {{}, {0xc000976010, 0x8}}}, 0xc000daba58, 0xc000b15ad0)
        /home/circleci/project/project/internal/terraform/eval_variable.go:140 +0xaed
github.com/hashicorp/terraform/internal/terraform.(*nodeModuleVariable).evalModuleVariable(0xc0006976e0?, {0x2b82d18?, 0xc000551420?}, 0xdd?)
        /home/circleci/project/project/internal/terraform/node_module_variable.go:239 +0x545
github.com/hashicorp/terraform/internal/terraform.(*nodeModuleVariable).Execute(0xc0006976e0, {0x2b82d18, 0xc000551420}, 0x4)
        /home/circleci/project/project/internal/terraform/node_module_variable.go:152 +0xd0
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc000b0e3c0, {0x2b82d18, 0xc000551420}, {0x7fcdb01aba78, 0xc0006976e0})
        /home/circleci/project/project/internal/terraform/graph_walk_context.go:136 +0xc2
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x23b1de0, 0xc0006976e0})
        /home/circleci/project/project/internal/terraform/graph.go:74 +0x2f0
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc000697740, {0x23b1de0, 0xc0006976e0}, 0xc00096aac0)
        /home/circleci/project/project/internal/dag/walk.go:381 +0x2f6
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update
        /home/circleci/project/project/internal/dag/walk.go:304 +0xf65
```
</details>

**Comment by liamcervante** [comment]
Hi @grimm26, thanks for the bug report! I can confirm I can replicate this, and that this worked in v1.3.3 and is now broken. I'm going to have a look into this to see if I can find a root cause and a plan for a fix.

@42Atomys, I think your issue is better marked as a duplicate of https://github.com/hashicorp/terraform/issues/32157, but without seeing the configuration I can't say for sure. If it is a duplicate of #32157, then it should already have a fix in place ready to be released in v1.3.5.

**Comment by 42Atomys** [comment]
Effectively thanks you 🙏

**Comment by liamcervante** [comment]
Okay, so I think the root cause of this issue is some upstream changes in the way go-cty handles optional attributes.

The end result of this is that when HCL attempts to work out the type for your conditional expression it fails because the objects within `lifecycle_presets` do not have matching attributes with each other and with the `lifecycle_object_rules`. 

A potential fix for this is to improve the [`UnifyUnsafe`](https://github.com/zclconf/go-cty/blob/main/cty/convert/public.go#L81) function within go-cty, such that it will allow unification between object types that do not have matching attributes by inserting additional concrete or optional attributes such that the resulting unified type matches both inputs. This is called from HCL [here](https://github.com/hashicorp/hcl/blob/main/hclsyntax/expression.go#L661).

As this fix is inside dependencies it may not make the v1.3.5 release window. A potential workaround for this use case is to make sure every object within `local.lifecycle_presents` has specified either null or a concrete value for every optional attribute. For example: https://gist.github.com/liamcervante/61c528c582bad058a7f293d70b75c398

I'll look into implementing the fix, hopefully that workaround can unblock you in the meantime.

**Comment by liamcervante** [comment]
Hi @grimm26, I've looked at this closer and we've discussed internally. Unfortunately we think this is a case of something that was a bug and shouldn't have been working previously, this means the behaviour now is actually correct. We fixed the root cause of the original bug (which was also causing several crashes) and hence this behaviour is no longer working.

It worked previously because the empty default list was retaining the optional attribute metadata when it shouldn't have, as optional attribute definitions are part of the go-cty type constraints but not part of concrete types. You can read a bit more about this in [this comment](https://github.com/zclconf/go-cty/pull/135#issuecomment-1273545511).

You can see this is an unintended side effect of the empty default list, because if you populate the default list of `lifecycle_object_rules` with anything concrete in v1.3.3 then your example would fail with the same error message as in v1.3.4. In that situation it is the empty default that is behaving badly rather than the populated default. The following example implementation of that variable will fail even in 1.3.3:

```hcl
variable "lifecycle_object_rules" {
  description = "List of object describing object lifecycle rules on the noncurrent versions of objects."
  type = list(object(
    {
      id                                     = string
      enabled                                = bool
      prefix                                 = optional(string)
      tags                                   = optional(map(string), {})
      object_size_greater_than               = optional(number)
      object_size_less_than                  = optional(number)
      abort_incomplete_multipart_upload_days = optional(number)
      transitions                            = optional(map(string), {})
      noncurrent_transitions                 = optional(map(string), {})
    }
  ))
  default = [
    {
      id = "my-test-id"
      enabled = false
    }
  ]
}
```

The correct thing for you to do here is the workaround described in my earlier comment. If you populate the local variables with null attributes so they all have every attribute defined then your implementation will start to work.

**Comment by liamcervante** [comment]
Working as intended.
